src/messaging/ExchangeMgr.cpp

```diff
@@ -273,7 +273,20 @@ void ExchangeManager::OnMessageReceived(const PacketHeader & packetHeader, const
     // handle the message.
     if (matchingUMH != nullptr || payloadHeader.NeedsAck())
     {
-        ExchangeDelegate * delegate = matchingUMH ? matchingUMH->Delegate : nullptr;
+        ExchangeDelegate * delegate = nullptr;
+
+        // Fetch delegate from the acceptor
+        if (matchingUMH != nullptr)
+        {
+            CHIP_ERROR err = matchingUMH->Acceptor->OnUnsolicitedMessageReceived(payloadHeader, msgBuf, delegate);
```

Comment:
The callee generally speaking has to allocate an ExchangeDelegate here.  Then code further on in this function will leak the delegate in various failure cases, starting with failure to allocate an exchange.  We need to set up the APIs so we can notify either the delegate or the acceptor (naming aside) that the delegate is in fact not needed.

Reply:
Added a `ReleaseDelegate` function to the UMH